dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/association/ProgramOrganisationUnitAssociationsQueryBuilder.java

```diff
@@ -0,0 +1,199 @@
+
+import org.hisp.dhis.commons.collection.CollectionUtils;
+import org.hisp.dhis.user.CurrentUserGroupInfo;
+import org.hisp.dhis.user.CurrentUserService;
+import org.hisp.dhis.user.User;
+import org.springframework.stereotype.Service;
+
+@Service
+@RequiredArgsConstructor
+public class ProgramOrganisationUnitAssociationsQueryBuilder
+{
+
+    private final CurrentUserService currentUserService;
+
+    private static final String SHARING_OUTER_QUERY_BEGIN = "select " +
+        "    prg.uid, " +
+        "    prg.agg_ou_uid " +
+        "from (";
+
+    private static final String SHARING_OUTER_QUERY_END = ") as prg";
+
+    private static final String INNER_SQL_QUERY = "select " +
+        "    pr.uid, " +
+        "    pr.sharing, " +
+        "    array_agg(ou.uid) agg_ou_uid " +
+        "from program pr " +
+        "    left outer join program_organisationunits po on pr.programid = po.programid " +
+        "    left outer join organisationunit ou on po.organisationunitid = ou.organisationunitid " +
+        "where";
+
+    private static final String INNER_QUERY_GROUPING_BY = "group by pr.uid, pr.sharing";
+
+    public String buildSqlQuery( Set<String> programUids, Set<String> userOrgUnitPaths, User currentUser )
+    {
+        Stream<String> queryParts = Stream.of(
+            SHARING_OUTER_QUERY_BEGIN,
+            innerQueryProvider( programUids, userOrgUnitPaths, currentUser ),
+            SHARING_OUTER_QUERY_END );
+
+        if ( nonSuperUser( currentUser ) )
+        {
+            queryParts = Stream.concat(
+                queryParts,
+                Stream.of(
+                    "where",
+                    getSharingConditions( LIKE_READ_METADATA ) ) );
+        }
+        return queryParts.collect( joining( " " ) );
+    }
+
+    private String innerQueryProvider( Set<String> programUids, Set<String> userOrgUnitPaths, User currentUser )
+    {
+        Stream<String> queryParts = Stream.of(
+            INNER_SQL_QUERY,
+            getProgramUidsFilter( programUids ) );
+
+        if ( nonSuperUser( currentUser ) )
+        {
+            queryParts = Stream.concat( queryParts,
+                Stream.of(
+                    "and",
+                    getUserOrgUnitPathsFilter( userOrgUnitPaths ) ) );
+        }
+
+        queryParts = Stream.concat( queryParts, Stream.of( INNER_QUERY_GROUPING_BY ) );
+
+        return queryParts.collect( joining( " " ) );
+    }
+
+    private String getSharingConditions( String access )
+    {
+        CurrentUserGroupInfo currentUserGroupInfo = currentUserService.getCurrentUserGroupsInfo();
+        return String.join( " or ",
+            getOwnerCondition( currentUserGroupInfo ),
+            getPublicSharingCondition( access ),
+            getUserGroupAccessCondition( currentUserGroupInfo, access ),
+            getUserAccessCondition( currentUserGroupInfo, access ) );
+    }
+
+    private String getOwnerCondition( CurrentUserGroupInfo currentUserGroupInfo )
```

Comment:
These methods look very generic. Maybe we have something for this already. Or they could be moved to a central location and re-used. @vietnguyen

Reply:
we have it for hibernate stores. This is its equivalent jdbc implementation. Maybe we can move to a separate class when needed